projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php

```diff
@@ -0,0 +1,249 @@
+		$this->task_registry[ $task['id'] ] = $task;
+		return true;
+	}
+
+	/**
+	 * Register a new Launchpad Task
+	 *
+	 * @param Task[] $tasks Collection of task definitions.
+	 *
+	 * @return bool True if successful, false if not.
+	 */
+	public function register_tasks( $tasks = array() ) {
+		$tasks_to_register = array();
+
+		foreach ( $tasks as $task ) {
+			// Register none of the tasks if any are invalid.
+			if ( ! $this->validate_task( $task ) ) {
+				return false;
+			}
+
+			$tasks_to_register[ $task['id'] ] = $task;
+		}
+
+		// TODO: Handle duplicate tasks
+		$this->task_registry = array_merge( $this->task_registry, $tasks_to_register );
+		return true;
+	}
+
+	/**
+	 * Unregister a Launchpad Task List
+	 *
+	 * @param string $id Task List id.
+	 *
+	 * @return bool True if successfully unregistered, false if not found.
+	 */
+	public function unregister_task_list( $id ) {
+		if ( ! array_key_exists( $id, $this->task_list_registry ) ) {
+			return false;
+		}
+
+		unset( $this->task_list_registry[ $id ] );
+		return true;
+	}
+
+	/**
+	 * Unregister a Launchpad Task
+	 *
+	 * @param string $id Task id.
+	 *
+	 * @return bool True if successful, false if not.
+	 */
+	public function unregister_task( $id ) {
+		if ( ! array_key_exists( $id, $this->task_registry ) ) {
+			return false;
+		}
+
+		unset( $this->task_registry[ $id ] );
+		return true;
+	}
+
+	/**
+	 * Get a Launchpad Task List definition
+	 *
+	 * @param string $id Task List id.
+	 *
+	 * @return Task_List Task List.
+	 */
+	protected function get_task_list( $id ) {
+		if ( ! array_key_exists( $id, $this->task_list_registry ) ) {
+			return array();
+		}
+
+		return $this->task_list_registry[ $id ];
+	}
+
+	/**
+	 * Get a Launchpad Task definition
+	 *
+	 * @param string $id Task id.
+	 *
+	 * @return Task Task.
+	 */
+	protected function get_task( $id ) {
+		if ( ! array_key_exists( $id, $this->task_registry ) ) {
+			return array();
+		}
+
+		return $this->task_registry[ $id ];
+	}
+
+	/**
+	 * Builds a collection of tasks for a given task list
+	 *
+	 * @param string $id Task list id.
+	 *
+	 * @return Task[] Collection of tasks associated with a task list.
+	 */
+	public function build( $id ) {
+		$task_list           = $this->get_task_list( $id );
+		$tasks_for_task_list = array();
+		// Takes a registered task list, looks at its associated task ids,
+		// and returns a collection of associated tasks.
+		foreach ( $task_list['task_ids'] as $task_id ) {
+			$tasks_for_task_list[] = $this->get_task( $task_id );
+		}
+
+		return $tasks_for_task_list;
+	}
+
+	/**
+	 * Validate a Launchpad Task List
+	 *
+	 * @param Task_List $task_list Task List.
+	 *
+	 * @return bool True if valid, false if not.
+	 */
+	public static function validate_task_list( $task_list ) {
+		if ( ! is_array( $task_list ) ) {
+			return false;
+		}
+
+		if ( ! isset( $task_list['id'] ) ) {
+			trigger_error( 'The Launchpad task list being registered requires a "id" attribute', E_USER_WARNING );
```

Comment:
We generally want stuff to be pretty lenient in WP(com) and not throw errors, at least never in production. There's too much stuff interacting weirdly across a large codebase like ours, so let's take a different approach. We do want devs to get feedback on how something's wrong, but not raise errors. A pretty good way is to use `_doing_it_wrong` which will send to the error log, and as its docs say:

> The current behavior is to trigger a user error if `WP_DEBUG` is true.

Which is a good way to run things on your sandbox.

Reply:
> We do want devs to get feedback on how something's wrong, but not raise errors. A pretty good way is to use _doing_it_wrong which will send to the error log, and as [its docs](https://developer.wordpress.org/reference/functions/_doing_it_wrong/) say:

Great feedback! This is exactly what I was trying to accomplish by setting the `trigger_error` error level to `E_USER_WARNING` ( doesn't fatal, but leaves a message in the error log )

Reply:
The `_doing_it_wrong` method doesn't seem to be sending anything to the error log, even when I set `WP_DEBUG` to true in the `wp-config` of my sandbox

Reply:
We can handle this in a follow up next sprint -- leaving a note in the tracking issue

Reply:
Noted